src/main/java/com/oath/halodb/HaloDBInternal.java

```diff
@@ -281,6 +286,66 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
         }
     }
 
+    public boolean takeSnapShot() {
+
+        logger.info("Start generating the snapshot");
+
+        File[] filesToCopy = dbDirectory.getPath().toFile()
+            .listFiles(file -> {
+                return (Constants.DATA_FILE_PATTERN.matcher(file.getName()).matches()
+                    || Constants.INDEX_FILE_PATTERN.matcher(file.getName()).matches()
+                    || Constants.TOMBSTONE_FILE_PATTERN.matcher(file.getName()).matches());
+            } );
+
+        logger.info("file number need to be copied: {}", filesToCopy.length);
+
+        try {
+            Path dbDirectoryPath = dbDirectory.getPath();
+            File snapShotDir = Paths.get(dbDirectoryPath.toFile().getAbsolutePath(), "snapshot").toFile();
+            // Only support one snapshot now
+            // TODO: support multiple snapshots if needed
+            if (snapShotDir.exists()) {
+                logger.warn("The snapshot dir is already existed. Delete the old one.");
+                FileUtils.deleteDirectory(snapShotDir);
+            }
+
+            FileUtils.createDirectoryIfNotExists(snapShotDir);
+            logger.info("Created directory for snapshot {}", snapShotDir.toString());
+
+            for (File file : filesToCopy) {
```

Comment:
If ```currentWriteFile``` is rolled over right after the file list finalized, the new file will be forcefully rolled over again.